src/uct/tcp/sockcm/sockcm_ep.c

```diff
@@ -14,26 +14,263 @@
+    recv_len = recv(ep->sock_id_ctx->sock_id, (char *) &conn_status,
+                    sizeof(conn_status), 0);
+    ucs_debug("sockcm_listener: recv len = %d\n", (int) recv_len);
+
+    status = conn_status ? UCS_ERR_REJECTED : UCS_OK;
+
+    pthread_mutex_lock(&ep->ops_mutex);
+    if (status == UCS_OK) {
+        ep->status = status;
+    } else {
+        iface = ucs_derived_of(ep->super.super.iface, uct_sockcm_iface_t);
+        uct_sockcm_ep_set_failed(&iface->super.super, &ep->super.super, status);
+    }
+    uct_sockcm_ep_invoke_completions(ep, status);
+    pthread_mutex_unlock(&ep->ops_mutex);
+}
+
 static UCS_CLASS_INIT_FUNC(uct_sockcm_ep_t, const uct_ep_params_t *params)
 {
-    uct_sockcm_iface_t *iface = ucs_derived_of(params->iface,
-                                               uct_sockcm_iface_t);
+    const ucs_sock_addr_t *sockaddr = params->sockaddr;
+    uct_sockcm_iface_t    *iface    = NULL;
+    char ip_port_str[UCS_SOCKADDR_STRING_LEN];
+    ucs_status_t status;
 
+    iface = ucs_derived_of(params->iface, uct_sockcm_iface_t);
     UCS_CLASS_CALL_SUPER_INIT(uct_base_ep_t, &iface->super);
+
+    if (iface->is_server) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (!(params->field_mask & UCT_EP_PARAM_FIELD_SOCKADDR)) {
+        return UCS_ERR_INVALID_PARAM;
+    }
+
+    UCT_SOCKCM_CB_FLAGS_CHECK((params->field_mask &
+			       UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+			      params->sockaddr_cb_flags : 0);
+
+    self->pack_cb       = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_PACK_CB) ?
+                          params->sockaddr_pack_cb : NULL;
+    self->pack_cb_arg   = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_USER_DATA) ?
+                          params->user_data : NULL;
+    self->pack_cb_flags = (params->field_mask &
+                           UCT_EP_PARAM_FIELD_SOCKADDR_CB_FLAGS) ?
+                          params->sockaddr_cb_flags : 0;
+    pthread_mutex_init(&self->ops_mutex, NULL);
+    ucs_queue_head_init(&self->ops);
+
+    if (sockaddr->addr->sa_family == AF_INET) {
+        memcpy(&self->remote_addr, sockaddr->addr, sizeof(struct sockaddr_in));
+    } else if (sockaddr->addr->sa_family == AF_INET6) {
+        memcpy(&self->remote_addr, sockaddr->addr, sizeof(struct sockaddr_in6));
+    } else {
+        ucs_error("sockcm ep: unknown remote sa_family=%d", sockaddr->addr->sa_family);
+        status = UCS_ERR_IO_ERROR;
+        goto err;
+    }
+
+    self->slow_prog_id = UCS_CALLBACKQ_ID_NULL;
+
+    status = uct_sockcm_ep_set_sock_id(iface, self);
+    if (status == UCS_ERR_NO_RESOURCE) {
+        goto add_to_pending;
+    } else if (status != UCS_OK) {
+        goto err;
+    }
+
+    self->is_on_pending = 0;
+
+    status = ucs_socket_connect(self->sock_id_ctx->sock_id, sockaddr->addr);
+    if (status != UCS_OK) {
+        goto err;
+    }
+
+    status = uct_sockcm_ep_send_client_info(iface, self);
```

Comment:
I think you need to move uct_sockcm_ep_send_client_info() to after setting the callback (line 185) to handle the message that was sent in uct_sockcm_ep_send_client_info()

Reply:
@alinask I handle the sent message inside uct_sockcm_ep_send_client_info (before I modify the socket as non-blocking) [here](https://github.com/openucx/ucx/pull/3570/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR76)

Are you suggesting this to handle the case when the sent message can't be completely sent when the send call returns? If yes, I could try and block after the send call to wait until all bytes are sent. The other alternative, as you suggest, is to or POLLOUT to the event handler and handle the sent message asynchronously and then modify the event handler to handle only POLLIN events from that point onwards. 

Let me know if it's because of other reasons.

Reply:
If you send the message first, and then set the function (uct_sockcm_ep_event_handler) to process it, can the sent message be lost?

Reply:
It shouldn't happen because `uct_sockcm_ep_send_client_info` should fail if the send returns successfully having written the expected number of bytes, correct? Is it possible that [this](https://github.com/openucx/ucx/pull/3570/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR76) assert succeeds but the message is still lost?

Reply:
Sorry my sentence is not clear!

`uct_sockcm_ep_send_client_info` will fail if the send doesn't send the expected number of bytes because of [this](https://github.com/openucx/ucx/pull/3570/files#diff-aca878e51b7d5e6eaa75cd20a1375eadR79) So I don't see the scenario when we set the handler and the message is lost.

With this logic, do we still need to make the event handler handle the send?

Reply:
doesn't seem so right now.